src/stores/tokenGovernanceContract.ts

```diff
@@ -0,0 +1,216 @@
+  abi?: ABIDefinition;
+  params?: EthAbiDecodeParametersResultObject;
+  raw: string;
+};
+
+type Proposal = {
+  num: number;
+  cancelled: boolean;
+  created: number;
+  effectiveDate: number;
+  msg: DecodedMessage;
+  subject;
+  methodStr: string;
+  summaryStr: string;
+  newValue?: string[];
+  currentValue?: string;
+};
+
+export class TokenGovernanceContractStore extends ContractStore {
+  private funcBySignature: Map<string, ABIDefinition>;
+
+  @observable
+  public noGovernance: boolean;
+  public props: any;
+  public state: { title: string; description: string };
+  public proposals: IObservableArray<Proposal>;
+
+  constructor() {
+    super(tokenGovernance, null);
+
+    this.state = {
+      title: '',
+      description: '',
+    };
+
+    this.props = {
+      title: '',
+      description: '',
+    };
+
+    reaction(
+      () => bridgeStore.contract,
+      () => {
+        bridgeStore.contract.methods
+          .admin()
+          .call()
+          .then(owner => {
+            this.address = owner;
+            this.funcBySignature = this.calculateAbiSignatures();
+            this.fetch();
+          });
+      }
+    );
+  }
+
+  private calculateAbiSignatures(): Map<string, ABIDefinition> {
+    return (tokenGovernance as ABIDefinition[])
+      .filter(m => m.type === 'function')
+      .reduce<ABIDefinitionBySig>(
+        (m: ABIDefinitionBySig, def: ABIDefinition) => {
+          const sig = web3RootStore.instance.eth.abi.encodeFunctionSignature(
+            def
+          );
+          m[sig] = def;
+          return m;
+        },
+        new Map()
+      );
+  }
+
+  private fetch() {
+    return Promise.all([
+      this.contract.methods.first().call(),
+      this.contract.methods.size().call(),
+      this.contract.methods.proposalTime().call(),
+    ])
+      .then(([first, size, proposalTime]) => {
+        return Promise.all(
+          (range(
+            Number(first),
+            Number(first) - 1 + Number(size)
+          ) as number[]).map(index => {
+            return this.contract.methods
+              .proposals(`${index}`)
+              .call()
+              .then(proposalData =>
+                this.enrichProposal(proposalData, index, proposalTime)
+              );
+          })
+        ).then(proposals => {
+          this.proposals = observable.array(proposals);
+        });
+      })
+      .catch(e => {
+        console.error('err', e);
+        // TODO: add better "no governance" detection
+        this.noGovernance = true;
+      });
+  }
+
+  public get tokenGovernanceContract(): Contract | undefined {
+    if (web3InjectedStore.instance) {
+      return new web3InjectedStore.instance.eth.Contract(
+        this.abi,
+        '0x3cc955f91d645b4250f6070a8b7d71365662776f'
+      );
+    }
+  }
+
+  @autobind
+  public async sendProposal() {
+    const accounts = await web3InjectedStore.instance.eth.getAccounts();
+
+    const proposalHashBuffer = keccak256(
```

Comment:
should upload title+description to IPFS, no?

Reply:
yep this is the goal later